File: frontend/src/lib/monitoring.ts
/**
 * Options for the Sentry SDK, which reports to our own Bugsink instance. Returns null when no
 * DSN is configured (local dev, tests), so the hooks skip initialisation entirely. Bugsink stores
 * errors only: tracing stays off, and no replay is loaded, which keeps the bundle small.
 */
export interface MonitoringEnv {
	dsn?: string;
	environment?: string;
	release?: string;
}

export interface SentryOptions {
	dsn: string;
	environment: string;
	release: string | undefined;
	tracesSampleRate: 0;
	sendDefaultPii: true;
}

export function sentryOptions({ dsn, environment, release }: MonitoringEnv): SentryOptions | null {
	if (!dsn) return null;
	return {
		dsn,
		environment: environment || 'development',
		release: release || undefined,
		tracesSampleRate: 0,
		sendDefaultPii: true
	};
}

File: backend/config/monitoring.py
"""Error reporting through the Sentry SDK to the self-hosted Bugsink instance.

Nothing is initialised unless SENTRY_DSN is set, so local development and tests never report.
Bugsink stores errors only, hence tracing is off; PII stays on because the instance is ours and
request context is what makes an error debuggable.
"""


def sentry_options(dsn: str, environment: str, release: str) -> dict | None:
    if not dsn:
        return None
    return {
        "dsn": dsn,
        "environment": environment,
        "release": release or None,
        "traces_sample_rate": 0,
        "send_default_pii": True,
    }


def init_sentry(env) -> bool:
    """Initialise the SDK from environment variables; returns whether it was enabled."""
    options = sentry_options(
        env("SENTRY_DSN", default=""),
        env("SENTRY_ENVIRONMENT", default="development"),
        env("APP_RELEASE", default=""),
    )
    if options is None:
        return False
    import sentry_sdk

    sentry_sdk.init(**options)
    return True
